prow/cmd/peribolos/main.go

```diff
@@ -244,34 +291,267 @@ func configureOrgMembers(opt options, client configureOrgMembersClient, orgName
 	if err := configureOrgMembers(opt, client, orgName, orgConfig); err != nil {
 		return fmt.Errorf("failed to configure %s members: %v", orgName, err)
 	}
 
-	// teams = set(all teams)
 	for name, team := range orgConfig.Teams {
-		team.Name = name
-		num := -1
-		// find team name in teams (or previous name)
-		if err := configureTeam(client, num, team); err != nil {
+		gt, ok := githubTeams[name]
+		if !ok { // configureTeams is buggy if this is the case
+			return fmt.Errorf("%s not found in id list", name)
+		}
+
+		// Configure team metadata
+		err := configureTeam(client, orgName, name, team, gt)
+		if err != nil {
 			return fmt.Errorf("failed to update %s: %v", orgName, err)
 		}
+
+		// Add/remove/update members to the team.
+		return configureTeamMembers(client, gt.ID, team)
 	}
 	return nil
 }
 
-func configureTeam(client *github.Client, num int, team org.Team) error {
-	// create team if num < 0
-	// else diff and patch
+type editTeamClient interface {
+	EditTeam(team github.Team) (*github.Team, error)
+}
+
+// configureTeam patches the team name/description/privacy when values differ
+func configureTeam(client editTeamClient, orgName, teamName string, team org.Team, gt github.Team) error {
+	// Do we need to reconfigure any team settings?
+	patch := false
+	if gt.Name != teamName {
+		patch = true
+	}
+	gt.Name = teamName
+	if team.Description != nil && gt.Description != *team.Description {
+		patch = true
+		gt.Description = *team.Description
+	} else {
+		gt.Description = ""
+	}
+	if team.Privacy != nil && gt.Privacy != string(*team.Privacy) {
+		patch = true
+		gt.Privacy = string(*team.Privacy)
+
+	} else {
+		gt.Privacy = ""
+	}
+
+	if patch { // yes we need to patch
+		if _, err := client.EditTeam(gt); err != nil {
+			return fmt.Errorf("failed to edit %s team %d(%s): %v", orgName, gt.ID, gt.Name, err)
+		}
+	}
+	return nil
+}
+
+// teamMembersClient can list/remove/update people to a team.
+type teamMembersClient interface {
+	ListTeamMembers(id int, role string) ([]github.TeamMember, error)
+	RemoveTeamMembership(id int, user string) error
+	UpdateTeamMembership(id int, user string, maintainer bool) (*github.TeamMembership, error)
+}
+
+// configureTeamMembers will add/update people to the appropriate role on the team, and remove anyone else.
+func configureTeamMembers(client teamMembersClient, id int, team org.Team) error {
+	// Get desired state
+	wantMaintainers := sets.NewString(team.Maintainers...)
+	wantMembers := sets.NewString(team.Members...)
+
+	// Get current state
+	haveMaintainers := sets.String{}
+	haveMembers := sets.String{}
+
+	members, err := client.ListTeamMembers(id, github.RoleMember)
+	if err != nil {
+		return fmt.Errorf("failed to list %d members", id)
+	}
+	for _, m := range members {
+		haveMembers.Insert(m.Login)
+	}
+
+	maintainers, err := client.ListTeamMembers(id, github.RoleMaintainer)
+	if err != nil {
+		return fmt.Errorf("failed to list %d maintainers", id)
```

Comment:
nit: we lose the actual `err` content here

Reply:
Done